### PR TITLE
ROS2 parameter library

### LINKED ISSUE #1529: [TODO] Parameter "library" with type safety for compile time checking
Pull request: #1527 

Link to TODO, or file number/line: `vision_receiver.cpp:29`

> I think at some point we should have a nicer interface for setting and getting parameters that keeps you from messing things up at compile time. Something like:
> 
> ```c++
> // vision_parameters.hpp
> struct SharedVisionPortSinglePrimary
> {
>     constexpr static std::string Name = "port";
>     using Type = int;
> };
> 
> // vision_receiver.cpp
> double hz = declare_and_get_parameter<SharedVisionPortSinglePrimary>();
> ```
> 
> (Probably with a macro to make the declaration side simpler)

_Originally posted by @kylestach in https://github.com/_render_node/MDE3OlB1bGxSZXF1ZXN0UmV2aWV3NDM0NTI3NjIw/pull_request_reviews/more_threads_

### PR DESCRIPTION
**Pull request by oswinso**
## Description
This PR creates a parameter library `rj_param_utils` for easy parameter declarations.

Closes #1529.

### Sim still works
1. Launch sim

Expected result: Everything still works.

### CODE REVIEW

File: rj_param_utils/include/rj_param_utils/param.h [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        : name_{name},
+          help_{help},
+          filename_{filename},
+          default_value_{param},
+          param_{param} {}
+
+    /**
+     * @brief Updates the parameter with the new value.
+     * @param new_value The new value of the parameter.
+     * @throws std::runtime_error if the type is set.
```
Comment (JNeiger) [inline rj_param_utils/include/rj_param_utils/param.h]: Maybe it's too early in the morning, but I don't think I get this comment?

File: rj_param_utils/include/rj_param_utils/param.h [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    template <typename ParamType>
+    ParamMap<ParamType>& GetParamMap();
+};
+
+}  // namespace params
+
+/**
+ * @brief Defines a namespaced parameter. The defined parameter can be used as
+ * PARAM_#, where * denotes # denotes the name passed in.
```
Comment (JNeiger) [inline rj_param_utils/include/rj_param_utils/param.h]: Typo? 
` * denotes # denotes`

File: rj_param_utils/include/rj_param_utils/param.h [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+ * @param type The type of the parameter.
+ * @param prefix The prefix / namespace of the parameter, using . as a
+ * separator.
+ * @param name The variable name of the parameter.
+ * @param val The default value of the parameter.
+ * @param description A description of the parameter.
+ */
+#define DEFINE_NAMESPACED_VARIABLE(type, prefix, name, val, description)   \
+    namespace params::variables {                                          \
+    static type PARAM_##name##_storage = val;                              \
```
Comment (JNeiger) [inline rj_param_utils/include/rj_param_utils/param.h]: Since only the name is being used here, does it conflict if you have two config values "config1/number" and "config2/number" as the prefix/name?

File: rj_param_utils/include/rj_param_utils/param.h [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    const std::string filename_;
+    const T& default_value_;
+    T& param_;
+};
+
+/**
+ * @brief A ParamProvider represents a "provider" for parameters. In particular,
+ * it should provide a way of updating the parameters.
+ */
+class ParamProvider {
```
Comment (JNeiger) [inline rj_param_utils/include/rj_param_utils/param.h:87]: Can put comments on the functions here since they're the main interface?

File: rj_param_utils/include/rj_param_utils/param.h [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+#define DEFINE_VARIABLE(type, name, val, description)          \
+    namespace params::variables {                              \
+    static type PARAM_##name##_storage = val;                  \
+    const type& PARAM_##name = PARAM_##name##_storage;         \
+    static ::params::internal::ParamRegisterer o_##name(       \
+        #name, description, __FILE__, PARAM_##name##_storage); \
+    }                                                          \
+    using params::variables::PARAM_##name;
+
+// Define the DEFINE_* macro for all supported types.
```
Comment (JNeiger) [inline rj_param_utils/include/rj_param_utils/param.h:191]: I'll let Kyle jump in, but I really like forcing the idea of namespaced parameters, it really keeps things from getting out of control in terms of keeping things clean. I would almost completely remove the root versions on just make the namespaced the default

File: rj_vision_receiver/src/vision_receiver.cpp [hunk 6]
```diff
@@ -9,17 +9,21 @@
...
+      port_{-1},
+      _socket{_io_context},
+      param_provider_(this) {
     _recv_buffer.resize(65536);
 
-    // Parameters. TODO(#1529): Better parameter "library".
-    declare_parameter<int>("port", SharedVisionPortSinglePrimary);
-    get_parameter<int>("port", port_);
-    setPort(port_);
+    setPort(PARAM_port);
```
Comment (JNeiger) [inline rj_vision_receiver/src/vision_receiver.cpp:28]: So this is set up in the "launch" files right? What would be the process to have two different configs ala sim and IRL and being able to swap between them in a simple way?

**Review by JNeiger — COMMENTED** [review]
Can you document more how the system works from a high level? The amount of indirection through templates/defines makes it hard to follow easily

File: rj_param_utils/src/ros2_param_provider.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    for (const auto& [param_name, param] : GetParamMap<type>()) { \
+        rcl_interfaces::msg::ParameterDescriptor descriptor;      \
+        descriptor.description = param->help();                   \
+        const type& val = node->declare_parameter(                \
+            param_name, param->default_value(), descriptor);      \
+        Update(param_name, val);                                  \
+    }
+
+    // Declare the parameters
+    DECLARE_AND_UPDATE_PARAMS(bool)
```
Comment (JNeiger) [inline rj_param_utils/src/ros2_param_provider.cpp:43]: These should be untabbed one

**Comment by oswinso** [comment]
Yeah, sure. I'll probably dump it in a google doc somewhere so that we can easily copy paste it when we do the documentation-a-thon week.

File: rj_vision_receiver/src/vision_receiver.cpp (hunk 6 above)
Comment (oswinso) [reply rj_vision_receiver/src/vision_receiver.cpp:28]: Yeah. When the `ROS2ParamProvider` is instantiated, that connects the `ROS2` parameter system with the parameters defined here.

If you wanted to have 2 different configs, you would need to create 2 different `.yaml` files, and then have some logic in the `.launch.py` file that chooses between the two.

File: rj_param_utils/include/rj_param_utils/param.h (hunk 3 above)
Comment (oswinso) [reply rj_param_utils/include/rj_param_utils/param.h]: Yeah, in that case it will conflict. I don't think it will conflict if they're in different translation units though.

Do you think this use case should be considered?

Comment (JNeiger) [reply rj_param_utils/include/rj_param_utils/param.h]: PID constants for different controllers is a good example. XY and Heading controllers both use a PID controller but with different gains

Comment (JNeiger) [reply rj_param_utils/include/rj_param_utils/param.h]: https://github.com/RoboJackets/robocup-software/blob/staging/soccer/config/sim.xml#L88
https://github.com/RoboJackets/robocup-software/blob/staging/soccer/config/sim.xml#L97

Comment (oswinso) [reply rj_param_utils/include/rj_param_utils/param.h]: I guess the easiest solution I can think of right now would be to put the `DEFINE_XXX` inside different namespaces on the client side.

Comment (JNeiger) [reply rj_param_utils/include/rj_param_utils/param.h]: Is it not possible to attach the prefix to all the stuff here? So like `PARAM_##prefix##__##name##_storage` or something like that. The prefix + name should be unique imo

Comment (oswinso) [reply rj_param_utils/include/rj_param_utils/param.h]: The problem with prefix is that ROS2 uses `.` as the separator for namespaces, ie. `a.b.c.example_param`, and `.` is an invalid character for C++ identifiers.

One way I could see this working would be to use `::` as the separator for the prefixes, ie.
`DEFINE_NAMESPACED_BOOL(a::b::c, thing, true, "description.")`, and then access it as `a::b::c::PARAM_thing` or something, and then do the conversion between `::` and `.` in `ROS2ParamProvider`.

File: rj_param_utils/include/rj_param_utils/param.h (hunk 1 above)
Comment (oswinso) [reply rj_param_utils/include/rj_param_utils/param.h]: yeah this comment is pretty retarded and doesn't make sense.

File: rj_param_utils/include/rj_param_utils/param.h (hunk 5 above)
Comment (oswinso) [reply rj_param_utils/include/rj_param_utils/param.h:191]: I think for simple nodes it does make sense to not need to have a namespace. ie. For the `port` parameter of `vision_receiver`

File: rj_param_utils/src/ros2_param_provider.cpp (hunk 7 above)
Comment (oswinso) [reply rj_param_utils/src/ros2_param_provider.cpp:43]: `clang-format` put these here so I'll leave them here.

File: rj_param_utils/include/rj_param_utils/param.h (hunk 5 above)
Comment (JNeiger) [reply rj_param_utils/include/rj_param_utils/param.h:191]: I would argue no namespaces is the exception to the rule. Unless a bunch is going to change, there aren't any non-namespace configs currently.

Unless I'm missing something, port as a parameter wouldn't be descriptive enough by itself in the context of the whole system. We would need to say vision_reciever_port or something, which is just a namespace tied into the name. I feel like I'm missing something though if that's the example.

Comment (oswinso) [reply rj_param_utils/include/rj_param_utils/param.h:191]: In ROS2, parameters are tied to the node, ie. the hierarchy is like
```
/vision_receiver
  - port
```

and so `port` would be in the "root namespace" under the `/vision_reciever` node.

Comment (JNeiger) [reply rj_param_utils/include/rj_param_utils/param.h:191]: That makes a lot more sense

File: rj_param_utils/include/rj_param_utils/param.h (hunk 2 above)
Comment (oswinso) [reply rj_param_utils/include/rj_param_utils/param.h]: Fixed.

File: rj_param_utils/include/rj_param_utils/param.h (hunk 4 above)
Comment (oswinso) [reply rj_param_utils/include/rj_param_utils/param.h:87]: Done.

**Review by JNeiger — APPROVED** [review]
👍

File: rj_param_utils/.clang-tidy [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+                  -fuchsia-default-arguments,
+                  -readability-redundant-member-init,
+                  -readability-magic-numbers,
+                  -readability-uppercase-literal-suffix,
+                  -misc-non-private-member-variables-in-classes,
+                  -cppcoreguidelines-non-private-member-variables-in-classes,
+                  -clang-analyzer-cplusplus.Move,
+                  -cppcoreguidelines-macro-usage,
+                  -bugprone-macro-parentheses
+                  '
```
Comment (kylestach) [inline rj_param_utils/.clang-tidy:28]: What is left out of this list?

File: rj_param_utils/include/rj_param_utils/param.h [hunk 9]
```diff
@@ -0,0 +1,233 @@
+/** @file */
+#pragma once
+
+#include <rclcpp/node.hpp>
+
+namespace params {
+namespace internal {
```
Comment (kylestach) [inline rj_param_utils/include/rj_param_utils/param.h:7]: nit: standard is `detail`, no?

**Review by kylestach — APPROVED** [review]
LGTM, at least for the time being.

File: rj_param_utils/.clang-tidy (hunk 8 above)
Comment (oswinso) [reply rj_param_utils/.clang-tidy:28]: More like `-cppcoreguidelines-macro-usage` and `-bugprone-macro-parentheses` were added.

File: rj_param_utils/include/rj_param_utils/param.h (hunk 9 above)
Comment (oswinso) [reply rj_param_utils/include/rj_param_utils/param.h:7]: what do you mean by standard is `detail`?
